### ISSUE TITLE
Linux - Switching MIDI devices leads to SIGABRT

### ISSUE THREAD

**Issue by RC2225** [luciusDXL/TheForceEngine#270]
Switching between MIDI Devices on Linux (EndeavourOS/Arch) while Dark Forces leads to a SIGABRT. It doesn't matter if music in the settings is enabled or disabled. It happens mostly after switching 2-3 times.

```
[Startup] TFE_Polygon::init
[Startup] TFE_Image::init
[Startup] TFE_FrontEndUI::init
[MemoryRegion] Allocated new memory block in region 'game' - new size is 1 blocks, total size is '8388608'
[MemoryRegion] Allocated new memory block in region 'level' - new size is 1 blocks, total size is '8388608'
[Progam Flow] The Force Engine Game Loop Started
[Game] Dark Forces Version: 1.0 (Build 1)
[MemoryRegion] Allocated new memory block in region 'Landru' - new size is 1 blocks, total size is '4194304'
[MemoryRegion] Allocated new memory block in region 'Cutscene' - new size is 1 blocks, total size is '8388608'
[Player] Setting up level 'DTENTION'                                                                                                                                                                            
[Warning : level_loadINF] Hit the end of INF 'DTENTION' before parsing all items: 107/200
[MemoryRegion] Allocated new memory block in region 'game' - new size is 2 blocks, total size is '16777216'
[Error : Midi Device] MidiOutAlsa::sendMessage: error sending MIDI message to port.
[Error : Midi Device] MidiOutAlsa::sendMessage: error sending MIDI message to port.
free(): invalid pointer

```

[trace.txt](https://github.com/luciusDXL/TheForceEngine/files/11578770/trace.txt)

**Comment by mlauss2** [comment]
I can reproduce; it's either a RtMidi usage bug, a bug in RtMidi itself, or in ALSA.
From a cursory glance at the backtrace I'm leaning towards a bug in RtMidi wrt. ALSA handling.

**Comment by mlauss2** [comment]
This here, applied on top of #278, fixes it for me.  When switching to/from midi devices, for a short time the main thread and the midiplayer thread both send midi messages, which, ultimately, causes ALSA to overwrite the glibc malloc housekeeping data (i.e. the 16 bytes in front of a pointer obtained by malloc) with garbage when a midi message is posted.  When the RtMidiOut device is destroyed, ALSA frees all its malloc()ed memories and glibc sanity checks kick in and abort() due to detected inconsistencies.

```diff
diff --git a/TheForceEngine/TFE_Audio/midiPlayer.cpp b/TheForceEngine/TFE_Audio/midiPlayer.cpp
index 3a3e5167..153ffb75 100644
--- a/TheForceEngine/TFE_Audio/midiPlayer.cpp
+++ b/TheForceEngine/TFE_Audio/midiPlayer.cpp
@@ -89,6 +89,13 @@ namespace TFE_MidiPlayer
 		TFE_System::logWrite(LOG_MSG, "Startup", "TFE_MidiPlayer::init");
 
 		bool res = false;
+
+		if (s_mutex) {
+			fprintf(stderr, "WHOA1\n");  // sollte nicht passieren...
+			delete s_mutex;
+		}
+		s_mutex = new std::mutex;
+
 		allocateMidiDevice(type);
 
 		if (s_midiDevice)
@@ -105,9 +112,6 @@ namespace TFE_MidiPlayer
 		}
 
 		s_runMusicThread.store(true);
-		if (s_mutex)
-			delete s_mutex;
-		s_mutex = new std::mutex;
 
 		s_thread = new std::thread(midiUpdateFunc, nullptr);
 
@@ -498,6 +503,7 @@ namespace TFE_MidiPlayer
 
 	void allocateMidiDevice(MidiDeviceType type)
 	{
+		s_mutex->lock();  // midi thread blockieren
 		delete s_midiDevice;
 		s_midiDevice = nullptr;
 
@@ -515,5 +521,6 @@ namespace TFE_MidiPlayer
 			default:
 				TFE_System::logWrite(LOG_ERROR, "Midi", "Invalid midi type selected: %d", (s32)type);
 		}
+		s_mutex->unlock();
 	}
 }
diff --git a/TheForceEngine/TFE_Audio/systemMidiDevice.cpp b/TheForceEngine/TFE_Audio/systemMidiDevice.cpp
index cf04d2e1..040d2ac2 100644
--- a/TheForceEngine/TFE_Audio/systemMidiDevice.cpp
+++ b/TheForceEngine/TFE_Audio/systemMidiDevice.cpp
@@ -10,6 +10,8 @@
 #include <TFE_System/system.h>
 #include <algorithm>
 
+#include <mutex>
+
 #ifdef _WIN32
 // Windows library required to access the midi device(s).
 #pragma comment( lib, "winmm.lib" )
@@ -26,15 +28,19 @@ namespace TFE_Audio
 		SMID_DRUM_CHANNEL = 9,
 	};
 
+	static std::mutex mtx;
+
 	void midiErrorCallback(RtMidiError::Type, const std::string&, void*);
 
 	SystemMidiDevice::SystemMidiDevice()
 	{
+		m_outputs.clear();
 		m_outputId = -1;
-		m_midiout = new RtMidiOut();
-		m_midiout->setErrorCallback(midiErrorCallback);
-
-		getOutputCount();
+		m_midiout = new RtMidiOut(RtMidi::UNSPECIFIED, "TFE");
+		if (m_midiout) {
+			m_midiout->setErrorCallback(midiErrorCallback);
+			getOutputCount();
+		}
 	}
 
 	SystemMidiDevice::~SystemMidiDevice()
@@ -44,14 +50,11 @@ namespace TFE_Audio
 
 	void SystemMidiDevice::exit()
 	{
-		if (m_outputId >= 0)
-		{
-			m_midiout->closePort();
-		}
+		m_outputId = -1;
+		m_midiout->closePort();
+		m_outputs.clear();
 		delete m_midiout;
 		m_midiout = nullptr;
-
-		m_outputId = -1;
 	}
 
 	const char* SystemMidiDevice::getName()
@@ -59,17 +62,22 @@ namespace TFE_Audio
 		return c_SystemMidi_Name;
 	}
 
+	void SystemMidiDevice::message(const u8* msg, u32 len)
+	{
+		unsigned int l = msg[0] == MID_PROGRAM_CHANGE ? 2 : 3;
+		if (m_outputId >= 0) {
+			mtx.lock();
+			m_midiout->sendMessage(msg, l);
+			mtx.unlock();
+		}
+	}
+
 	void SystemMidiDevice::message(u8 type, u8 arg1, u8 arg2)
 	{
 		const u8 msg[3] = { type, arg1, arg2 };
-		if (m_midiout) { m_midiout->sendMessage(msg, 3); }
+		message(&msg[0], 3);
 	}
 	
-	void SystemMidiDevice::message(const u8* msg, u32 len)
-	{
-		if (m_midiout) { m_midiout->sendMessage(msg, (size_t)len); }
-	}
-
 	void SystemMidiDevice::noteAllOff()
 	{
 		for (u32 c = 0; c < MIDI_CHANNEL_COUNT; c++)
@@ -85,7 +93,7 @@ namespace TFE_Audio
 
 	u32 SystemMidiDevice::getOutputCount()
 	{
-		if (m_outputs.empty())
+		if (m_outputs.empty() && m_midiout)
 		{
 			u32 count = m_midiout->getPortCount();
 			m_outputs.resize(count);
```

**Comment by mlauss2** [comment]
Here's a more generalized fix, on top of #278 .  It reuses the lock in the midi thread to serialize writes to the midi port, and gets rid of a few RtMidiOut error messages and annoyances in systemmididevice.

```diff
diff --git a/TheForceEngine/TFE_Audio/midiPlayer.cpp b/TheForceEngine/TFE_Audio/midiPlayer.cpp
index 3a3e5167..9167d0c0 100644
--- a/TheForceEngine/TFE_Audio/midiPlayer.cpp
+++ b/TheForceEngine/TFE_Audio/midiPlayer.cpp
@@ -89,6 +89,10 @@ namespace TFE_MidiPlayer
 		TFE_System::logWrite(LOG_MSG, "Startup", "TFE_MidiPlayer::init");
 
 		bool res = false;
+		s_mutex = new std::mutex();
+		if (!s_mutex)
+			return false;
+
 		allocateMidiDevice(type);
 
 		if (s_midiDevice)
@@ -105,9 +109,6 @@ namespace TFE_MidiPlayer
 		}
 
 		s_runMusicThread.store(true);
-		if (s_mutex)
-			delete s_mutex;
-		s_mutex = new std::mutex;
 
 		s_thread = new std::thread(midiUpdateFunc, nullptr);
 
@@ -322,7 +323,7 @@ namespace TFE_MidiPlayer
 		}
 	}
 
-	void stopAllNotes()
+	static void _stopAllNotes()
 	{
 		// Some devices don't support "all notes off" - so do it manually.
 		for (s32 i = 0; i < MIDI_INSTRUMENT_COUNT; i++)
@@ -350,7 +351,14 @@ namespace TFE_MidiPlayer
 		memset(s_instrOn, 0, sizeof(Instrument) * MIDI_INSTRUMENT_COUNT);
 		s_curNoteTime = 0.0;
 	}
-		
+
+	void stopAllNotes()
+	{
+		s_mutex->lock();	// block midi thread
+		_stopAllNotes();
+		s_mutex->unlock();	// unblock midi thread
+	}
+
 	void sendMessageDirect(u8 type, u8 arg1, u8 arg2)
 	{
 		u8 msg[] = { type, arg1, arg2 };
@@ -365,7 +373,12 @@ namespace TFE_MidiPlayer
 			s_channelSrcVolume[channelIndex] = arg2;
 			msg[2] = u8(s_channelSrcVolume[channelIndex] * s_masterVolumeScaled);
 		}
-		if (s_midiDevice) { s_midiDevice->message(msg, len); }
+		if (s_midiDevice)
+		{
+			s_mutex->lock();		// block midi thread
+			s_midiDevice->message(msg, len);
+			s_mutex->unlock();		// unblock midi thread
+		}
 
 		// Record currently playing instruments and the note-on times.
 		if (msgType == MID_NOTE_OFF || msgType == MID_NOTE_ON)
@@ -399,7 +412,8 @@ namespace TFE_MidiPlayer
 				if ((s_instrOn[i].channelMask & channelMask) && (s_curNoteTime - s_instrOn[i].time[c] > s_maxNoteLength))
 				{
 					// Turn off the note.
-					sendMessageDirect(MID_NOTE_OFF | c, i);
+					if (s_midiDevice)
+						s_midiDevice->message(MID_NOTE_OFF | c, i);
 
 					// Reset the instrument channel information.
 					s_instrOn[i].channelMask &= ~channelMask;
@@ -434,7 +448,7 @@ namespace TFE_MidiPlayer
 					{
 						localTimeCallback = 0;
 						isPaused = true;
-						stopAllNotes();
+						_stopAllNotes();
 					} break;
 					case MIDI_RESUME:
 					{
@@ -448,7 +462,7 @@ namespace TFE_MidiPlayer
 					} break;
 					case MIDI_STOP_NOTES:
 					{
-						stopAllNotes();
+						_stopAllNotes();
 						// Reset callback time.
 						localTimeCallback = 0;
 						s_midiCallback.accumulator = 0.0;
@@ -463,7 +477,9 @@ namespace TFE_MidiPlayer
 				s_midiCallback.accumulator += TFE_System::updateThreadLocal(&localTimeCallback);
 				while (s_midiCallback.callback && s_midiCallback.accumulator >= s_midiCallback.timeStep)
 				{
+					s_mutex->unlock();	// callback might call a locking function
 					s_midiCallback.callback();
+					s_mutex->lock();
 					s_midiCallback.accumulator -= s_midiCallback.timeStep;
 					s_curNoteTime += s_midiCallback.timeStep;
 				}
@@ -498,6 +514,7 @@ namespace TFE_MidiPlayer
 
 	void allocateMidiDevice(MidiDeviceType type)
 	{
+		s_mutex->lock();	// block midi thread
 		delete s_midiDevice;
 		s_midiDevice = nullptr;
 
@@ -515,5 +532,6 @@ namespace TFE_MidiPlayer
 			default:
 				TFE_System::logWrite(LOG_ERROR, "Midi", "Invalid midi type selected: %d", (s32)type);
 		}
+		s_mutex->unlock();	// unblock midi thread
 	}
 }
diff --git a/TheForceEngine/TFE_Audio/systemMidiDevice.cpp b/TheForceEngine/TFE_Audio/systemMidiDevice.cpp
index cf04d2e1..07a1ddd7 100644
--- a/TheForceEngine/TFE_Audio/systemMidiDevice.cpp
+++ b/TheForceEngine/TFE_Audio/systemMidiDevice.cpp
@@ -31,6 +31,7 @@ namespace TFE_Audio
 	SystemMidiDevice::SystemMidiDevice()
 	{
 		m_outputId = -1;
+		m_outputs.clear();
 		m_midiout = new RtMidiOut();
 		m_midiout->setErrorCallback(midiErrorCallback);
 
@@ -46,11 +47,13 @@ namespace TFE_Audio
 	{
 		if (m_outputId >= 0)
 		{
+			noteAllOff();
 			m_midiout->closePort();
 		}
 		delete m_midiout;
 		m_midiout = nullptr;
 
+		m_outputs.clear();
 		m_outputId = -1;
 	}
 
@@ -62,12 +65,12 @@ namespace TFE_Audio
 	void SystemMidiDevice::message(u8 type, u8 arg1, u8 arg2)
 	{
 		const u8 msg[3] = { type, arg1, arg2 };
-		if (m_midiout) { m_midiout->sendMessage(msg, 3); }
+		if (m_outputId >= 0) { m_midiout->sendMessage(msg, 3); }
 	}
 	
 	void SystemMidiDevice::message(const u8* msg, u32 len)
 	{
-		if (m_midiout) { m_midiout->sendMessage(msg, (size_t)len); }
+		if (m_outputId >= 0) { m_midiout->sendMessage(msg, (size_t)len); }
 	}
 
 	void SystemMidiDevice::noteAllOff()
@@ -85,7 +88,7 @@ namespace TFE_Audio
 
 	u32 SystemMidiDevice::getOutputCount()
 	{
-		if (m_outputs.empty())
+		if (m_outputs.empty() && m_midiout)
 		{
 			u32 count = m_midiout->getPortCount();
 			m_outputs.resize(count);
@@ -100,7 +103,7 @@ namespace TFE_Audio
 
 	void SystemMidiDevice::getOutputName(s32 index, char* buffer, u32 maxLength)
 	{
-		if (index < 0 || index >= (s32)getOutputCount()) { return; }
+		if (index < 0 || index >= (s32)getOutputCount() || !m_midiout) { return; }
 
 		const std::string& name = m_outputs[index];
 		const u32 copyLength = std::min((u32)name.length(), maxLength - 1);
@@ -110,7 +113,6 @@ namespace TFE_Audio
 
 	bool SystemMidiDevice::selectOutput(s32 index)
 	{
-		if (!m_midiout) { return false; }
 		if (index < 0 || index >= (s32)getOutputCount())
 		{
 			index = 0;
```

**Comment by luciusDXL** [comment]
This should be fixed by the latest SDL 2 Mutex change.
